File: medium/3418-maximum-amount-of-money-robot-can-earn/solution.py
from typing import List


class Solution:
    def maximumAmount(self, coins: List[List[int]]) -> int:
        """
        3418. Maximum Amount of Money Robot Can Earn
        Time: O(m × n)
        Space: O(m × n)
        """
        m, n = len(coins), len(coins[0])

        dp = [[[float('-inf')] * 3 for _ in range(n)] for _ in range(m)]

        if coins[0][0] >= 0:
            for k in range(3):
                dp[0][0][k] = coins[0][0]
        else:
            dp[0][0][0] = coins[0][0]
            dp[0][0][1] = 0
            dp[0][0][2] = 0

        for i in range(m):
            for j in range(n):
                if i == 0 and j == 0:
                    continue

                for k in range(3):
                    best_prev = float('-inf')

                    if i > 0:
                        best_prev = max(best_prev, dp[i - 1][j][k])
                    if j > 0:
                        best_prev = max(best_prev, dp[i][j - 1][k])

                    if best_prev == float('-inf'):
                        continue

                    val = coins[i][j]

                    dp[i][j][k] = max(dp[i][j][k], best_prev + val)

                    if val < 0 and k > 0:
                        best_prev_k_minus = float('-inf')
                        if i > 0:
                            best_prev_k_minus = max(best_prev_k_minus, dp[i - 1][j][k - 1])
                        if j > 0:
                            best_prev_k_minus = max(best_prev_k_minus, dp[i][j - 1][k - 1])

                        dp[i][j][k] = max(dp[i][j][k], best_prev_k_minus)

        return max(dp[m - 1][n - 1])

File: medium/3418-maximum-amount-of-money-robot-can-earn/solution.java
/**
 * 3418. Maximum Amount of Money Robot Can Earn
 * Time: O(m × n)
 * Space: O(m × n)
 */
class Solution {
    public int maximumAmount(int[][] coins) {
        int m = coins.length, n = coins[0].length;

        int[][][] dp = new int[m][n][3];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < 3; k++) {
                    dp[i][j][k] = Integer.MIN_VALUE / 2;
                }
            }
        }

        if (coins[0][0] >= 0) {
            for (int k = 0; k < 3; k++) {
                dp[0][0][k] = coins[0][0];
            }
        } else {
            dp[0][0][0] = coins[0][0];
            dp[0][0][1] = 0;
            dp[0][0][2] = 0;
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 && j == 0) continue;

                for (int k = 0; k < 3; k++) {
                    int val = coins[i][j];

                    if (i > 0) {
                        dp[i][j][k] = Math.max(dp[i][j][k], dp[i - 1][j][k] + val);
                    }

                    if (j > 0) {
                        dp[i][j][k] = Math.max(dp[i][j][k], dp[i][j - 1][k] + val);
                    }

                    if (val < 0 && k > 0) {
                        if (i > 0) {
                            dp[i][j][k] = Math.max(dp[i][j][k], dp[i - 1][j][k - 1]);
                        }
                        if (j > 0) {
                            dp[i][j][k] = Math.max(dp[i][j][k], dp[i][j - 1][k - 1]);
                        }
                    }
                }
            }
        }

        return Math.max(dp[m - 1][n - 1][0],
                Math.max(dp[m - 1][n - 1][1], dp[m - 1][n - 1][2]));
    }
}
